lib/logstash/inputs/elasticsearch.rb

```diff
@@ -160,38 +157,42 @@ def run(output_queue)
       r = process_next_scroll(output_queue, r['_scroll_id'])
       has_hits = r['has_hits']
     end
-  end # def run
+  end
 
   private
+
   def process_next_scroll(output_queue, scroll_id)
     r = scroll_request(scroll_id)
     r['hits']['hits'].each { |hit| push_hit(hit, output_queue) }
     {'has_hits' => r['hits']['hits'].any?, '_scroll_id' => r['_scroll_id']}
   end
 
-  private
   def push_hit(hit, output_queue)
     event = LogStash::Event.new(hit['_source'])
     decorate(event)
 
     if @docinfo
-      event[@docinfo_target] ||= {}
+      # do not assume event[@docinfo_target] to be in-place updatable. first get it, update it, then at the end set it in the event.
+      docinfo_target = event[@docinfo_target] || {}
 
-      unless event[@docinfo_target].is_a?(Hash)
-        @logger.error("Elasticsearch Input: Incompatible Event, incompatible type for the `@metadata` field in the `_source` document, expected a hash got:", :metadata_type => event[@docinfo_target].class)
+      unless docinfo_target.is_a?(Hash)
```

Comment:
Given that the only API we require from whatever `docinfo_target` is `[]=` maybe that is a better test.  Also in line 176 we are quite comfortable with making it a new Hash if it is nil consider this:
```ruby
docinfo_target = event[@docinfo_target]
unless docinfo_target.respond_to?("[]=")
  @logger.warn("Elasticsearch Input: Incompatible type for the docinfo_target=#{@docinfo_target} field in the `_source` document, using a hash instead of:", :docinfo_target_type => docinfo_target.class, :event => event)
  docinfo_target = {}
end
```
For me, this means we try our utmost to ingest.
WDYT?